File: CombinationSumII.java
class Solution {
    private void helper(List<List<Integer>> rez, int target, ArrayList<Integer> current, int[] candidates, int start) {
        if(target == 0) {
            rez.add(new ArrayList<>(current));
        }

        for(int i = start; i<candidates.length; i++) {
            if(i > start && candidates[i-1] == candidates[i]) {
                continue;
            }
            if(target < 0) {
                return;
            }
            
            current.add(candidates[i]);
            helper(rez, target - candidates[i], current, candidates, i+1);
            current.remove(current.size()-1);
        }
    }
    
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);

        List<List<Integer>> rez = new ArrayList<>();
        ArrayList<Integer> current = new ArrayList<>();
        helper(rez, target, current, candidates, 0);
        return rez;
    }
}

File: CombinationSumII.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        std::vector<std::vector<int>> rez{};
        std::vector<int> combination;
        std::sort(candidates.begin(), candidates.end()); 

        findCombinations(candidates, target, 0, combination, rez);
        return rez;
    }

private:
    void findCombinations(const vector<int>& candidates, int target, int start, vector<int>& combination, vector<vector<int>>& result) {
        if (target == 0) {
            result.push_back(combination);
            return;
        }
        
        for (int i = start; i < candidates.size(); ++i) {
            if (i > start && candidates[i] == candidates[i - 1]) {
                continue; // Skip duplicates
            }
            
            if (candidates[i] > target) {
                break;
            }
            
            combination.push_back(candidates[i]);
            findCombinations(candidates, target - candidates[i], i + 1, combination, result); // Move to the next index
            combination.pop_back(); // Backtrack
        }
    }
};
